certora/D3MHub.spec

```diff
@@ -0,0 +1,1376 @@
+    assert(
+        tic == 0 && active && // regular path in normal path
+        targetAssets >= assetsBefore && // plan determines we need to go up (or keep the same)
+        targetAssets > lineWad && // target IS restricted by ilk line
+        (LineBefore - debtMiddle) / RAY() >= targetAssets - assetsBefore && // target IS NOT restricted by global Line
+        inkBefore > lineWad && // ink before execution is not safe (over ilk line)
+        maxWithdraw < assetsBefore - inkBefore // NOT enough for full rebalance
+            => artAfter == inkBefore &&
+               assetsAfter > artAfter,
+               "unwind: error 5"
+    );
+    assert(
+        tic == 0 && active && // regular path in normal path
+        targetAssets >= assetsBefore && // plan determines we need to go up (or keep the same)
+        targetAssets > lineWad && // target IS restricted by ilk line
+        (LineBefore - debtMiddle) / RAY() >= targetAssets - assetsBefore && // target IS NOT restricted by global Line
+        inkBefore > lineWad && // ink before execution is not safe (over ilk line)
+        maxWithdraw < inkBefore - lineWad // no way to decrease to ilk line value
+            => artAfter == inkBefore + fixInk - maxWithdraw &&
+               assetsAfter >= artAfter,
+               "unwind: error 6"
+    );
+    assert(
+        tic == 0 && active && // regular path in normal path
+        targetAssets >= assetsBefore && // plan determines we need to go up (or keep the same)
+        targetAssets > lineWad && // target IS restricted by ilk line
+        (LineBefore - debtMiddle) / RAY() >= targetAssets - assetsBefore && // target IS NOT restricted by global Line
+        inkBefore > lineWad && // ink before execution is not safe (over ilk line)
+        maxWithdraw < assetsBefore - lineWad && // NOT enough to rebalance and decrease to ilk line value
+        maxWithdraw >= assetsBefore - inkBefore // enough for full rebalance
+            => artAfter == assetsBefore - maxWithdraw &&
+               artAfter <= inkBefore &&
+               artAfter >= lineWad &&
+               assetsAfter == artAfter,
+               "unwind: error 7"
+    );
+    // Force unwinding due to ilk caged (but not culled yet) or plan inactive:
+    assert(
+        (tic > 0 || !active) &&
+        assetsBefore == maxWithdraw
+            => artAfter == 0, "unwind: error 8"
+    );
+    assert(
+        (tic > 0 || !active) &&
+        assetsBefore - maxWithdraw < lineWad
+            => artAfter == assetsBefore - maxWithdraw, "unwind: error 9"
+    );
+    assert(
+        (tic > 0 || !active) &&
+        assetsBefore - maxWithdraw >= lineWad &&
+        inkBefore <= lineWad
+            => artAfter == lineWad, "unwind: error 10"
+    );
+    assert(
+        (tic > 0 || !active) &&
+        assetsBefore - maxWithdraw >= inkBefore &&
+        inkBefore > lineWad
+            => artAfter == inkBefore, "unwind: error 11"
+    );
+}
+
+rule exec_normal_revert(bytes32 ilk) {
```

Comment:
probably need to rename `normal` to `_exec`

Reply:
Yeah not sure what is the best as `exec_exec` will clash with the other spec which calls `exec(ilk)` twice.
Also it is more a symbolic name, as we don't have `vat_caged` or `ilk_culled` neither.